libraries/ms-common/src/x86/soft_timer.c

```diff
@@ -1,58 +1,111 @@
+// Timers are implemented as a doubly linked list, sorted in order of remaining
+// time until expiry. Thus, the head of the list is always defined as the next
+// timer to expire, so we back the list with a linux interval timer that's always
+// set to the head's remaining time.
 #include "soft_timer.h"
 
 #include <signal.h>
-#include <stdbool.h>
-#include <stddef.h>
-#include <stdint.h>
-#include <stdio.h>
+#include <string.h>
 #include <time.h>
 
 #include "critical_section.h"
 #include "interrupt_def.h"
+#include "log.h"
+#include "objpool.h"
 #include "status.h"
 #include "x86_interrupt.h"
 
-typedef struct POSIXTimer {
-  timer_t timer_id;
+#define SOFT_TIMER_GET_ID(timer) ((timer)-s_storage)
+
+typedef struct SoftTimer {
+  uint32_t remainder;
   SoftTimerCallback callback;
   void *context;
-  volatile bool inuse;
-  bool created;
-} POSIXTimer;
-
-static struct sigevent s_event;
-static POSIXTimer s_posix_timers[SOFT_TIMER_MAX_TIMERS];
-static volatile uint8_t s_active_timers = 0;
+  struct SoftTimer *next;
+  struct SoftTimer *prev;
+} SoftTimer;
+
+typedef struct SoftTimerList {
+  SoftTimer *head;
+  ObjectPool pool;
+  timer_t timer_id;  // Single linux timer to back our soft timers
+  bool created;      // bool for tracking timer creation in case of re-initialization
+  struct sigevent event;
+} SoftTimerList;
+
+static SoftTimerList s_list = { 0 };
+
+// Backing array for timer list objpool
+static volatile SoftTimer s_storage[SOFT_TIMER_MAX_TIMERS] = { 0 };
+
+// Set the linux timer to a new value. Checks for divide by zero.
+static void prv_set_timer(uint32_t val_us) {
+  // See timer_settime manual page for details
+  struct itimerspec spec = {
+    { 0, 0 },                                                                 //
+    { val_us ? val_us / 1000000 : 0, val_us ? val_us % 1000000 * 1000 : 0 },  //
```

Comment:
I don't think there's any divide-by-zero danger here lol, the ternaries seem redundant.

Reply:
lol im dumb, thanks for that